s3_start_transfer/src/s3_start_transfer.py

```diff
@@ -0,0 +1,141 @@
+    import requests
+from os import fsencode
+import urllib.parse
+
+
+
+def am_api_post_json(url, data):
+    """
+    POST json to the Archivematica API
+    :param api_path: URL path to request (without hostname, e.g. /api/v2/location/)
+    :param data: Dict of data to post
+    :returns dict of json data returned by request:
+    """
+    am_url = os.environ['ARCHIVEMATICA_URL']
+    am_user = os.environ['ARCHIVEMATICA_USERNAME']
+    am_api_key = os.environ['ARCHIVEMATICA_API_KEY']
+    am_headers = {"Authorization": "ApiKey {0}:{1}".format(am_user, am_api_key)}
+
+    url = "{0}{1}".format(am_url, url)
+    print("URL: %s; Data: %s" % (url, data))
+    response = requests.post(url, json=data, headers=am_headers)
+    print("Response: %s" % response)
+    response_json = response.json()
+    print("Response JSON: %s" % response_json)
+    return response_json
+
+
+def ss_api_get(api_path, params=None):
+    """
+    GET request to the Archivematica storage service API
+    :param api_path: URL path to request (without hostname, e.g. /api/v2/location/)
+    :param params: Dict of params to include in the request
+    :returns dict of json data returned by request:
+    """
+    ss_url = os.environ['ARCHIVEMATICA_SS_URL']
+    ss_user = os.environ['ARCHIVEMATICA_SS_USERNAME']
+    ss_api_key = os.environ['ARCHIVEMATICA_SS_API_KEY']
+    ss_headers = {"Authorization": "ApiKey {0}:{1}".format(ss_user, ss_api_key)}
+
+    params = params or {}
+    url = "{0}{1}".format(ss_url, api_path)
+    print("URL: %s; Params: %s" % (url, params))
+    response = requests.get(url, params=params, headers=ss_headers)
+    print("Response: %s" % response)
+    response_json = response.json()
+    print("Response JSON: %s" % response_json)
+    return response_json
+
+
+def get_target_location(bucket, key):
+    """
+    Get an S3 location from the Archivematica storage service
+    to match the given bucket and key
+
+    :param bucket: Name of s2 bucket
+    :param key: Name of s3 key
+    :returns: bytestring identifying location, of the form
+        b'<location_uuid>:<target_path>'
+    """
+    # Get s3 location matching bucket and key
+    # Look for an S3 transfer source that matches the bucket name and prefix
+    # and take the transfer ID of that location
+    relative_path = ''
+    target_path = ''
+    ts_location_uuid = None
+    s3_sources = ss_api_get(
+        '/api/v2/location/',
+        {
+            'space__access_protocol': 'S3',
+            'purpose': 'TS'
+        }
+    )
+    for location in s3_sources['objects']:
+        relative_path = location['relative_path'].strip(os.sep)
+
+        if key.startswith(relative_path):
+            space_bucket = ss_api_get(location['space'])['s3_bucket']
+            if space_bucket == bucket:
+                ts_location_uuid = location['uuid']
+                target_path = key.split(relative_path, 1)[-1]
+
+                return fsencode(ts_location_uuid) + b":" + fsencode(target_path)
```

Comment:
It’d be useful to have an example of the Archivematica API output here, so we can read and check this code without a working Archivematica installation to hand. And a quick note about what a “location” is in this context – is it somewhere Archivematica knows to write files, or a list of all the files it’s ever written.

I don't think we need to initialise the `relative_path`, `target_path` and `ts_location_uuid` variables, do we?

What if there isn't a matching file in the Archivematica API? It would be good to throw a specific exception instead of returning `None` and falling over somewhere else.

If you pull out the core logic into a separate function, it might be easier to test? Something like:

```python
s3_sources = ss_api_get(…)

all_spaces = set(location["space"] for location in s3_sources["objects"])
buckets = {space: ss_api_get(space)["s3_bucket"] for space in all_spaces}

find_matching_location(s3_sources, buckets, bucket, key)
```

Reply:
Thanks for the suggestions! I've made some improvements to those functions, tidied up a bit and added some comments - as well as adding some exceptions :)